File: docs/migration/lib/utility.js
const fs = require('fs');
const getDirName = require('path').dirname;
const path = require('path');

function* readAllFilesFromDirectory(dir) {
  const files = fs.readdirSync(dir, { withFileTypes: true });

  for (const file of files) {
    if (file.isDirectory()) {
      yield* readAllFilesFromDirectory(path.join(dir, file.name));
    } else {
      yield path.join(dir, file.name);
    }
  }
}

function writeFile(path, contents) {
  fs.mkdir(getDirName(path), { recursive: true}, function (err) {
    if (err) return;

    fs.writeFileSync(path, contents, 'utf8');
  });
}

function findFrontmatterEndIndex(markdown) {
  const frontmatterPattern = /^---\s*[\s\S]*?---\s*/m;
  const match = frontmatterPattern.exec(markdown);
  if (match) {
    return match.index + match[0].length;
  }
  return 0;
}

function toPascalCase(string) {
  return `${string}`
    .toLowerCase()
    .replace(new RegExp(/[-_]+/, 'g'), ' ')
    .replace(new RegExp(/[^\w\s]/, 'g'), '')
    .replace(
      new RegExp(/\s+(.)(\w*)/, 'g'),
      ($1, $2, $3) => `${$2.toUpperCase() + $3}`
    )
    .replace(new RegExp(/\w/), s => s.toUpperCase());
}

function test(name, value, expectation) {
  if (value !== expectation) {
    throw new Error(`❌ ${name} failed. Expecting:\n${expectation} but instead got:\n${value}`)
  }

  console.log(`✅ ${name} test passed`)
}

module.exports = {
  readAllFilesFromDirectory,
  writeFile,
  findFrontmatterEndIndex,
  toPascalCase,
  test
};

File: docs/migrationv2/utility.go
package main

import (
	"regexp"
	"strings"
)

// TODO: use the standard libray in place of this
// function* readAllFilesFromDirectory(dir) {
//   const files = fs.readdirSync(dir, { withFileTypes: true });
//
//   for (const file of files) {
//     if (file.isDirectory()) {
//       yield* readAllFilesFromDirectory(path.join(dir, file.name));
//     } else {
//       yield path.join(dir, file.name);
//     }
//   }
// }

// TODO: use the standard library in place of this
// function writeFile(path, contents) {
//   fs.mkdir(getDirName(path), { recursive: true}, function (err) {
//     if (err) return;
//
//     fs.writeFileSync(path, contents, 'utf8');
//   });
// }

func findFrontmatterEndIndex(markdown string) int {
	frontmatterPattern := regexp.MustCompile(`^---\s*[\s\S]*?---\s*`)
	loc := frontmatterPattern.FindStringIndex(markdown)
	if loc == nil {
		return 0
	}
	return loc[1]
}

func toPascalCase(in string) string {
	sep := regexp.MustCompile("[-_]+")
	w := regexp.MustCompile(`\s+(.)`)
	s := strings.ToLower(in)
	s = sep.ReplaceAllString(s, " ")
	s = strings.TrimLeft(s, `\w\s`)
	return w.ReplaceAllStringFunc(s, strings.ToUpper)
}
